### PR TITLE
KAFKA-5921: add Materialized overloads to windowed kstream

### PR DESCRIPTION
**Pull request by dguy**
Add `Materialized` overloads to `WindowedKStream`. Deprecate existing methods on `KGroupedStream`

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @bbejeck @mjsax

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/WindowedKStream.java [hunk 1]
```diff
@@ -71,6 +73,45 @@
     KTable<Windowed<K>, Long> count();
 
     /**
+     * Count the number of records in this stream by the grouped key and the defined windows.
+     * Records with {@code null} key or value are ignored.
+     * The specified {@code windows} define either hopping time windows that can be overlapping or tumbling (c.f.
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/WindowedKStream.java]: This is not introduced by this PR: we do not have a parameter `windows` anymore in these functions, so I'm wondering if we should extract this part of the explanation on window specs to the javadoc of the `WindowedKStream` class itself?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java [hunk 2]
```diff
@@ -235,7 +223,7 @@ private void determineIsQueryable(final String queryableStoreName) {
                                                                   final Aggregator<? super K, ? super V, T> aggregator,
                                                                   final Windows<W> windows,
                                                                   final Serde<T> aggValueSerde) {
-        return windowedBy(windows).aggregate(initializer, aggregator, aggValueSerde);
+        return windowedBy(windows).aggregate(initializer, aggregator);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: Should we really ignore the user passed `aggValueSerde`, even if it is a deprecated function? I think to achieve correctness we should still wrap it in `Materialized`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedKStreamImpl.java [hunk 3]
```diff
@@ -60,7 +63,7 @@
 
     @Override
     public KTable<Windowed<K>, Long> count() {
-        return aggregate(
+        return doAggregate(
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedKStreamImpl.java]: We can use `countInitializer` and `countAggregator` below as well?

**Review by guozhangwang — COMMENTED** [review]
Minor comments, otherwise LGTM.

**Comment by dguy** [comment]
Thanks @guozhangwang, i've updated with your feedback if you want to have one more look

**Comment by guozhangwang** [comment]
Thanks @dguy , LGTM.

**Comment by dguy** [comment]
merged to trunk

**Pull request closed without merging**